types/datum.go

```diff
@@ -878,7 +878,8 @@ func ProduceFloatWithSpecifiedTp(f float64, target *FieldType, sc *stmtctx.State
 			return f, errors.Trace(err)
 		}
 	}
-	if mysql.HasUnsignedFlag(target.Flag) && f < 0 {
+	if mysql.HasUnsignedFlag(target.Flag) && f < 0 && sc.ShouldClipToZero() {
+		//                                         ^^^^^^^^^^^^^^^^^^^^^^^^ fix issue #16028
```

Comment:
Maybe we can delete this line for clean code.

Reply:
Besides, can we add more test cases when `sc.ShouldClipToZero() == true`  and `f < 0`? Because after my [investigation](https://github.com/pingcap/tidb/issues/16028#issuecomment-608371283), I found that there was only [handle](https://github.com/pingcap/tidb/blob/54e7ace1e02028cc6be1dff59208231ee432663c/table/column.go#L176-L181) the `truncation` error. And the `overflow` error will be [handled incorrectly](https://github.com/pingcap/tidb/blob/54e7ace1e02028cc6be1dff59208231ee432663c/table/column.go#L182-L184), and I hope to fix this problem in this PR. What's your opinion? @sduzh

Reply:
No problem, I will add more test cases. @Reminiscent